Add impact metric

AIceste opened the pull request:
- Recommend squash and merging.
- Implemented as a feature you can switch on and off with
cmake -DSUPPORT_VAR_IMPACT=1
 so no impact unless desired

- - - - - - - - 

I did this as a school project under Ph.D. Claude-Guy Quimper, at ULaval, Quebec, QC, Canada.
Please don't hold him responsible for any wrong done by this PR.

Whilst doubtful in practical value in real-world use case, the metric still shows pretty good results in some use cases from the minizinc benchmark collection (https://github.com/MiniZinc/minizinc-benchmarks). 

And more importantly, I believe it has a significant value for academic purpose.
I believe Chuffed is still used a lot as primary solver in introductory-level courses.

As things stand, impact variable selection is still a rare occurrence in the field, and as a student it can be hard to see how it can truly affect search, and to learn its various caveats / benefits.

This leads to a hole in knowledge, and more importantly in interest in this heuristic, where we could expect great things from future work.

AIceste committed:
Add switches to enable and use Impact heuristic

AIceste committed:
Add impact metric and scoring - still unused

Mostly a checkpoint before I start burning code

AIceste committed:
Add impact processing to the engine

AIceste committed:
Learn to read properly, adjust default impact

AIceste committed:
Remove extra debug include

AIceste committed:
Add resolution impact evaluation

schutta commented:
The impact search would be a welcome extension of Chuffed, before a merge we need to review your code.

schutta commented on chuffed/vars/int-var.h:
```diff
@@ -251,7 +263,11 @@ class IntVar : public Var {
 
 	int size() const {
+#ifdef HAS_VAR_IMPACT
+		if (!vals) return max - min;
```
Why did you use "max - min" and not the "count" below?

AIceste replied:
I wanted the option to be entirely independent from other options.

From my understanding, the current implementation of size can be called only if INT_DOMAIN_LIST, or if 'vals' is confirmed to be used, and will crash otherwise.

For instance, in IntVar::getScore, VAR_SIZE min and max return size based on max - min calculations, unless vals is defined.

The commented code is now outdated from fixing the off-by-one error.
As a side note, I would add that benchmark actually did not notice any slowing-down from adding the +1, so we could consider making VAR_SIZE scoring consistent.

schutta commented:
I had a quick look, it seems for me that the impact variable selection is always chosen when "SUPPORT_VAR_IMPACT" is on, i.e., no other variable selection can be chosen. If this is correct then it is a sub-optimal solution and you should change the implementation in this way that it becomes an option, i.e., not using macros to switch it on and off.

AIceste committed:
Fix off-by-one & typo

AIceste committed:
Set constant parameters

AIceste commented:
> I had a quick look, it seems for me that the impact variable selection is always chosen when "SUPPORT_VAR_IMPACT" is on, i.e., no other variable selection can be chosen. If this is correct then it is a sub-optimal solution and you should change the implementation in this way that it becomes an option, i.e., not using macros to switch it on and off.

It might be a lack of understanding from my part, but I believe I updated the switch cases properly, and the impact variable selection should only occur when the proper VarBranch value is given to Branching. Also, benchmarking my implementation against the original Chuffed always gave the exact same results (visited node count, the process time had slight variations, usually taking more time, as expected).

The impact value of integer variables, however, is always calculated (if SUPPORT_VAR_IMPACT), regardless of the chosen branching. 

As I stated in some french report I joined to my work:

A prior implementation was limiting impact calculation to variables that were identified to branch on impact, but in my opinion
- this could end up being a limiting factor when implementing "dynamic branching", which could normalise score values and compare various metrics (VarBranch score) against each other.
- if someone wants to compare impacts of his variables, or fetch statitics thorough the search, this is currently easy to do. 


I would like to add that the current constants chosen for the specific implementation I'm proposing have been chosen from a selection, but could still use some benchmarking with a wider variety of problems and values. I have a (ugly) script I used to generate hundreds of binaries, and another one used to benchmark them minimally if git let me I'll insert its last version here (it should be ran in cmake build directory) (the version I'm sending is minimal, but all arrays could be much longer) :

`# run in cmake build directory, first argument is output directory

import subprocess
import sys

weights = [
	[0, 0, 42],
]
cleareds = [
	[0, 0.42, 0.42],
]

dampings = [-0.26, 0.16, 0.58]

for d in dampings:
	with open('../chuffed/vars/int-var.h', 'r+') as f:
		f.write(f"""#ifndef int_var_h
#define int_var_h

#include <new>
#include <stdint.h>
#include <chuffed/support/misc.h>
#include <chuffed/core/sat-types.h>
#include <chuffed/core/engine.h>
#include <chuffed/vars/vars.h>
#include <chuffed/vars/bool-view.h>

#ifndef INT64_MAX
#define INT64_MAX 9223372036854775807LL
#endif
#ifndef INT64_MIN
#define INT64_MIN (-INT64_MAX-1)
#endif

// Controls whether active domain values are stored in a trailed linked list,
// to allow fast enumeration of domain values.  Also whether a count is kept.
// NOTE:  Makes search a bit slower, so kept it conditional at this stage,
// turning this off selects compatibility routines that scan value-by-value.
#define INT_DOMAIN_LIST 0

#define IMPACT_DAMPING {d:5.2f}""")
	for c in cleareds:
		for w in weights:
			with open('../chuffed/branching/impact.cpp', 'r+') as f:
				f.write(f"""#include <chuffed/branching/impact.h>

constexpr double WEIGHT_LOCAL = {w[0]:3d};
constexpr double WEIGHT_GLOBAL = {w[1]:3d};
constexpr double WEIGHT_REPARTITION = {w[2]:3d};
constexpr double CLEARED_LOCAL = {c[0]:4.2f};
constexpr double CLEARED_GLOBAL = {c[1]:4.2f};
constexpr double CLEARED_REPARTITION = {c[2]:4.2f};""")
			subprocess.run(['cmake', '--build', '.'])	
			subprocess.run(['mv', 'fzn-chuffed', f'{sys.argv[1]}/{w[0]}_{w[1]}_{w[2]}_{c[0]}_{c[1]}_{c[2]}_{d}'])
`

To generate basic data : 
`
import subprocess
import sys

weights = [
	[0, 0, 42],
]
cleareds = [
	[0.42, 0.42, 1],
]

dampings = [-0.26, 0.16, 0.58]

files = [
	'amaze_impact.fzn',
	'grid-colouring_impact.fzn',
	'shortest_path_impact.fzn',
]

args = [
	['-v', '-f', '-t420000'],
	['-v', '-t420000'],
]
test_count = 3

tests = [{
		'bin': f'./fzn_chuffed/{w[0]}_{w[1]}_{w[2]}_{c[0]}_{c[1]}_{c[2]}_{d}',
		'file': f, 'args': a,
		'weights': w, 'cleareds': c, 'damping': d,
		'durations': [], 'nodes': []
	} 
	for d in dampings
	for c in cleareds
	for w in weights
	for a in args
	for f in files
]
for _ in range(test_count):
	for test in tests:
		sorties = (subprocess.run([test['bin']] + test['args'] + [test['file']], text=True,
									stdout=subprocess.PIPE).stdout).split('%%%mzn-stat: ')
		test['durations'] += [float((sorties[15].split('solveTime='))[-1])]
		test['nodes'] += [int((sorties[1].split('nodes='))[-1])]
		with open(sys.argv[1], 'a') as f:
			f.write(str(test) + '\n')

	avgs = [{
			'bin': t['bin'], 'file': t['file'], 'args': t['args'],
			'weights': t['weights'], 'cleareds': t['cleareds'],
			'damping': t['damping'],
			'duration': sum(t['durations'])/len(t['durations']),
			'nodes': sum(t['nodes']) / len(t['nodes'])
		} for t in tests
	]

	c_stats = []
	for c in cleareds:
		for a in args:
			for f in files:
				tmp = [avg
					for avg in avgs
					if avg['file'] == f and avg['args'] == a and avg['cleareds'] == c
				]
				c_stats.append({
					'cleareds': c, 'file': f, 'args': a,
					'duration': sum([avg['duration'] for avg in tmp]) / len(tmp),
					'nodes': sum([avg['nodes'] for avg in tmp]) / len(tmp)
				})
			
	d_stats = []
	for d in dampings:
		for a in args:
			for f in files:
				tmp = [avg
					for avg in avgs
					if avg['file'] == f and avg['args'] == a and avg['damping'] == d
				]
				d_stats.append({
					'damping': d, 'file': f, 'args': a,
					'duration': sum([avg['duration'] for avg in tmp]) / len(tmp),
					'nodes': sum([avg['nodes'] for avg in tmp]) / len(tmp)
				})

	w_stats = []
	for w in weights:
		for a in args:
			for f in files:
				tmp = [avg
					for avg in avgs
					if avg['file'] == f and avg['args'] == a and avg['weights'] == w
				]
				d_stats.append({
					'weights': w, 'file': f, 'args': a,
					'duration': sum([avg['duration'] for avg in tmp]) / len(tmp),
					'nodes': sum([avg['nodes'] for avg in tmp]) / len(tmp)
				})
	rw_stats = []
	wnames = ['Wlocal', 'Wglobal', 'Wrep']
	for i in range(3):
		for a in args:
			for f in files:
			        # those should be updated to reflect weight combinations.. sorry about that I did quick and dirty
				for rw in [0, 42/882, 42/504, 42/126, 42/84, 42/42, 420/882, 420/504, 420/462]:
					try:
						tmp = [avg
							for avg in avgs
							if avg['file'] == f and avg['args'] == a and rw == avg['weights'][i] / sum(avg['weights'])
						]
						rw_stats.append({
							wnames[i]: rw, 'file': f, 'args': a,
							'duration': sum([avg['duration'] for avg in tmp]) / len(tmp),
							'nodes': sum([avg['nodes'] for avg in tmp]) / len(tmp)
						})
					except:
						pass
				
			

	with open(sys.argv[1], 'a') as f:
		f.write('\n======================================\n')
		for test in avgs:
			f.write(f"{test['bin']} {' '.join(test['args'])} {test['file']} AVERAGE: {test['duration']}s, {round(test['nodes'])} nodes\n")
		f.write('\n\n')
		for test in c_stats:
			f.write(f'{test}\n')
		f.write('\n\n')
		for test in d_stats:
			f.write(f'{test}\n')
		f.write('\n\n')
		for test in w_stats:
			f.write(f'{test}\n')
		f.write('\n\n')
		for test in rw_stats:
			f.write(f'{test}\n')

		f.write('\n======================================\n')
`

AIceste committed:
Remove debug code that should not serve a purpose anymore

AIceste commented:
Yes hello, I've been cleaning the public face of this GitHub account.

After answering @schutta 's comment in April, I have not heard from anyone.
Is there an action that I can still take to satisfy merge requirements?

I also went and cut down the verbosity of my comments, I'll add a separate comment mentioning the fine-tuning of literals in the impact formula.

AIceste commented:
**Fine tuning IMPACT formula variables**

At the beginning of [branching/impact.cpp]( https://github.com/chuffed/chuffed/pull/73/files#diff-608b1248990fdc53f53c87278603d9e4185fecdc72d00177c631a293e56dc87c), one can see six constexpr values that are parameters to the impact calculation formula.

When working on this, I spent a few dozen CPU hours benchmarking different values against some of the more well known minizinc classics, and whilst there were some definitive losers, there really weren't any "definitive winner".

Since the purpose is mainly academic, I don't think it is a great issue that the impact is not always optimised to fit the greater number of applications ; like all other heuristics, its use will depend on the problem at hand, and too bad if it isn't a fit.

Nonetheless, some might want to fine-tune these values for their specific application, and even if I didn't want to include my scripts to the chuffed repo (they don't belong here), I think having them *somewhere* is important.

Here is a sample of what I used to benchmark different values.

**bench.py**
```py
import subprocess
import sys

weights = [
	[42, 0, 42],
	[42, 0, 420],
	[42, 42, 42],
	[42, 42, 420],
	[42, 420, 42],
	[42, 420, 420],
	[420, 0, 42],
	[420, 42, 42],
	[420, 42, 420],
	[420, 420, 42],
]
cleareds = [
	[0.42, 0.042, 1],
]

dampings = [-0.26, 0.16, 0.58]

files = [
	'amaze_impact.fzn',
	'grid-colouring_impact.fzn',
	'shortest_path_impact.fzn',
]

args = [
	['-v', '-f', '-t42000'],
	['-v', '-t42000'],
]
test_count = 2

tests = [{
		'bin': f'./fzn_chuffed/{w[0]}_{w[1]}_{w[2]}_{c[0]}_{c[1]}_{c[2]}_{d}',
		'file': f, 'args': a,
		'weights': w, 'cleareds': c, 'damping': d,
		'durations': [], 'nodes': []
	} 
	for d in dampings
	for c in cleareds
	for w in weights
	for a in args
	for f in files
]
for _ in range(test_count):
	for test in tests:
		sorties = (subprocess.run([test['bin']] + test['args'] + [test['file']], text=True,
									stdout=subprocess.PIPE).stdout).split('%%%mzn-stat: ')
		test['durations'] += [float((sorties[15].split('solveTime='))[-1])]
		test['nodes'] += [int((sorties[1].split('nodes='))[-1])]
		with open(sys.argv[1], 'a') as f:
			f.write(str(test) + '\n')

	avgs = [{
			'bin': t['bin'], 'file': t['file'], 'args': t['args'],
			'weights': t['weights'], 'cleareds': t['cleareds'],
			'damping': t['damping'],
			'duration': sum(t['durations'])/len(t['durations']),
			'nodes': sum(t['nodes']) / len(t['nodes'])
		} for t in tests
	]

	c_stats = []
	for c in cleareds:
		for a in args:
			for f in files:
				tmp = [avg
					for avg in avgs
					if avg['file'] == f and avg['args'] == a and avg['cleareds'] == c
				]
				c_stats.append({
					'cleareds': c, 'file': f, 'args': a,
					'duration': sum([avg['duration'] for avg in tmp]) / len(tmp),
					'nodes': sum([avg['nodes'] for avg in tmp]) / len(tmp)
				})
			
	d_stats = []
	for d in dampings:
		for a in args:
			for f in files:
				tmp = [avg
					for avg in avgs
					if avg['file'] == f and avg['args'] == a and avg['damping'] == d
				]
				d_stats.append({
					'damping': d, 'file': f, 'args': a,
					'duration': sum([avg['duration'] for avg in tmp]) / len(tmp),
					'nodes': sum([avg['nodes'] for avg in tmp]) / len(tmp)
				})

	w_stats = []
	for w in weights:
		for a in args:
			for f in files:
				tmp = [avg
					for avg in avgs
					if avg['file'] == f and avg['args'] == a and avg['weights'] == w
				]
				d_stats.append({
					'weights': w, 'file': f, 'args': a,
					'duration': sum([avg['duration'] for avg in tmp]) / len(tmp),
					'nodes': sum([avg['nodes'] for avg in tmp]) / len(tmp)
				})
	rw_stats = []
	wnames = ['Wlocal', 'Wglobal', 'Wrep']
	for i in range(3):
		for a in args:
			for f in files:
				for rw in [0, 42/882, 42/504, 42/126, 42/84, 42/42, 420/882, 420/504, 420/462]:
					try:
						tmp = [avg
							for avg in avgs
							if avg['file'] == f and avg['args'] == a and rw == avg['weights'][i] / sum(avg['weights'])
						]
						rw_stats.append({
							wnames[i]: rw, 'file': f, 'args': a,
							'duration': sum([avg['duration'] for avg in tmp]) / len(tmp),
							'nodes': sum([avg['nodes'] for avg in tmp]) / len(tmp)
						})
					except:
						pass
				
			

	with open(sys.argv[1], 'a') as f:
		f.write('\n======================================\n')
		for test in avgs:
			f.write(f"{test['bin']} {' '.join(test['args'])} {test['file']} AVERAGE: {test['duration']}s, {round(test['nodes'])} nodes\n")
		f.write('\n\n')
		for test in c_stats:
			f.write(f'{test}\n')
		f.write('\n\n')
		for test in d_stats:
			f.write(f'{test}\n')
		f.write('\n\n')
		for test in w_stats:
			f.write(f'{test}\n')
		f.write('\n\n')
		for test in rw_stats:
			f.write(f'{test}\n')

		f.write('\n======================================\n')
		


```

**build_impact_params.py**
```py
# run in cmake build directory, first argument is output directory

import subprocess
import sys

weights = [
	[0, 0, 42],
	[0, 42, 0],
	[0, 42, 42],
	[0, 42, 420],
	[0, 420, 42],
	[42, 0, 0],
	[42, 0, 42],
	[42, 0, 420],
	[42, 42, 0],
	[42, 42, 42],
	[42, 42, 420],
	[42, 420, 0],
	[42, 420, 42],
	[42, 420, 420],
	[420, 0, 42],
	[420, 42, 0],
	[420, 42, 42],
	[420, 42, 420],
	[420, 420, 42],
]
cleareds = [
	[0, 0.42, 0.42],
	[0.42, 0.42, 0.42],
	[1, 0.42, 0.42],
	[0.42, 0, 0.42],
	[0.42, 1, 0.42],
	[0.42, 0.42, 0],
	[0.42, 0.42, 1],
]

dampings = [-0.26, 0.16, 0.58]

for d in dampings:
	with open('../chuffed/vars/int-var.h', 'r+') as f:
		f.write(f"""#ifndef int_var_h
#define int_var_h

#include <new>
#include <stdint.h>
#include <chuffed/support/misc.h>
#include <chuffed/core/sat-types.h>
#include <chuffed/core/engine.h>
#include <chuffed/vars/vars.h>
#include <chuffed/vars/bool-view.h>

#ifndef INT64_MAX
#define INT64_MAX 9223372036854775807LL
#endif
#ifndef INT64_MIN
#define INT64_MIN (-INT64_MAX-1)
#endif

// Controls whether active domain values are stored in a trailed linked list,
// to allow fast enumeration of domain values.  Also whether a count is kept.
// NOTE:  Makes search a bit slower, so kept it conditional at this stage,
// turning this off selects compatibility routines that scan value-by-value.
#define INT_DOMAIN_LIST 0

#define IMPACT_DAMPING {d:5.2f}""")
	for c in cleareds:
		for w in weights:
			with open('../chuffed/branching/impact.cpp', 'r+') as f:
				f.write(f"""#include <chuffed/branching/impact.h>

constexpr double WEIGHT_LOCAL = {w[0]:3d};
constexpr double WEIGHT_GLOBAL = {w[1]:3d};
constexpr double WEIGHT_REPARTITION = {w[2]:3d};
constexpr double CLEARED_LOCAL = {c[0]:4.2f};
constexpr double CLEARED_GLOBAL = {c[1]:4.2f};
constexpr double CLEARED_REPARTITION = {c[2]:4.2f};""")
			subprocess.run(['cmake', '--build', '.'])	
			subprocess.run(['mv', 'fzn-chuffed', f'{sys.argv[1]}/{w[0]}_{w[1]}_{w[2]}_{c[0]}_{c[1]}_{c[2]}_{d}'])

```

schutta commented on chuffed/branching/impact.cpp:
```diff
@@ -0,0 +1,70 @@
+constexpr double WEIGHT_REPARTITION = 420;
+constexpr double CLEARED_LOCAL = 0.42;
+constexpr double CLEARED_GLOBAL = 0.042;
+constexpr double CLEARED_REPARTITION = 1.00;;
```
Please remove double-colon.

schutta committed:
Update impact.cpp

Removed double-colon.
